src/treequest/vis/build_snapshot.py

```diff
@@ -16,17 +18,63 @@
 StateT = TypeVar("StateT")
 AlgoStateT = TypeVar("AlgoStateT")
 
+try:  # Optional dependency – used only if available
+    from pydantic import BaseModel as PydanticBaseModel  # type: ignore
+except Exception:  # pragma: no cover - pydantic is optional
+    PydanticBaseModel = None
+
+
+def _default_state_formatter(state: Any) -> str:
+    """Default formatter for node states.
+
+    - If the state is a pydantic BaseModel, prefer JSON.
+    - If the state is a dataclass instance, serialize via asdict() to JSON.
+    - Fallback to repr()/str() otherwise.
+    """
+    # Pydantic BaseModel → JSON
+    if PydanticBaseModel is not None and isinstance(state, PydanticBaseModel):
+        try:
+            if hasattr(state, "model_dump_json"):
+                # Pydantic v2 preferred API
+                return state.model_dump_json()
+            if hasattr(state, "model_dump"):
+                # Pydantic v2 Python object → JSON string
+                return json.dumps(state.model_dump(), default=str)
+            if hasattr(state, "json"):
+                # Pydantic v1 API
+                return state.json()
+        except Exception:
+            # Fall through to generic formatting
+            pass
+
+    # Dataclass instance → JSON
+    if dataclasses.is_dataclass(state) and not isinstance(state, type):
+        try:
+            return json.dumps(dataclasses.asdict(state), default=str)
```

Comment:
Same reason as above.
```suggestion
            return json.dumps(dataclasses.asdict(state), default=str, separators=(",", ":"))
```